src/core/time.ts

```diff
@@ -118,6 +118,9 @@ export class TimeValue {
    * @param other The time value to add to this one.
    */
   add(other: TimeValue): TimeValue {
+    if (other.isNever()) {
+      return this;
+    }
```

Comment:
This is an interesting definition of adding Never to the TimeValue. However, I think we should be cautious when adding semantics to an operation and see if this definition makes sense. Is this definition used in other targets as well? How about checking if the other time value is Never before applying the add operation and using the time as is rather than defining an addition using never?

Reply:
Formally, I think `never` is -:infinity:, but here it seems to be treated as zero, which I don't think is quite right. What prompted this change, @byeong-gil?

Reply:
In reactor-c, when summing a `tag_t` and an `interval_t`, the `NEVER` interval is treated as 'no delay'. And the 'ZERO' interval adds a microstep.

Here, the `lf_delay_tag` function returns the parameter `tag` as is when the parameter `interval` is `NEVER`.
https://github.com/lf-lang/reactor-c/blob/907ba0e86a3cf7956e96257ab51c2d89c3c71069/core/tag.c#L133-L151

Here, `NEVER` is used to indicate 'no delay'.
https://github.com/lf-lang/lingua-franca/blob/0864dd0287932f994fb3a56e2bc9870d5c63b619/core/src/main/java/org/lflang/federated/extensions/CExtensionUtils.java#L457

Also, I wanted to unify the meaning of these two functions 
https://github.com/lf-lang/reactor-ts/blob/14f5b1352723185a1c8d5834168883ad380f332c/src/core/time.ts#L441-L447
https://github.com/lf-lang/reactor-ts/blob/14f5b1352723185a1c8d5834168883ad380f332c/src/core/time.ts#L120-L130

Reply:
I think this is ok, but let's wait for @lhstrh 's opinions.

Reply:
If @lhstrh agrees with this change, How about adding a microstep if the interval is `TimeValue.zero()` to align the function `getLaterTag` with [`lf_delay_tag`](https://github.com/lf-lang/reactor-c/blob/907ba0e86a3cf7956e96257ab51c2d89c3c71069/core/tag.c#L133-L151) in `reactor-c`?
i.e.
```
getLaterTag(delay: TimeValue): Tag { 
  if (this.time.isNever() || delay.isNever()) { 
    return this; 
  } else if (delay.isZero() {
    return new Tag(this.time, this.microstep + 1);
  } else { 
    return new Tag(delay.add(this.time), 0); 
  } 
} 
```

Reply:
Hmm.. I'm frankly a bit puzzled by the examples in the C runtime/code generator that @byeong-gil is sharing. Treating `NEVER` as "no delay", for instance, doesn't sound right to me. Perhaps @edwardalee can shed some light on this.

Reply:
The reason we treat NEVER as no delay is that we have to make a distinction between zero delay and no delay. interval_t is an int64, and there is only one encoding for zero.  I think the reasoning was that an interval of minus infinity doesn't seem very useful. It is the time between a point in time and the time something never happened (?) ... a bit of a brain twister... Like trying to figure out when the big bang happened (hint: it happened NEVER, as you can determine by working time backwards and applying relativity as gravitation increases).

Reply:
> If @lhstrh agrees with this change, How about adding a microstep if the interval is `TimeValue.zero()` to align the function `getLaterTag` with [`lf_delay_tag`](https://github.com/lf-lang/reactor-c/blob/907ba0e86a3cf7956e96257ab51c2d89c3c71069/core/tag.c#L133-L151) in `reactor-c`? i.e.
> 
> ```
> getLaterTag(delay: TimeValue): Tag { 
>   if (this.time.isNever() || delay.isNever()) { 
>     return this; 
>   } else if (delay.isZero() {
>     return new Tag(this.time, this.microstep + 1);
>   } else { 
>     return new Tag(delay.add(this.time), 0); 
>   } 
> } 
> ```

This looks OK to me, except I think that we should handle the "forever" case specially as well, because if we don't, that would lead to overflow. Concretely, I think that we should have :infinity: + something = :infinity:.

Reply:
One thing that confused me was that `NEVER` was defined as `0x8000000000000000` in reactor-c (that is something like negative infinity in `int64`???) - I guess it's just a practical choice. I need to bang me head several times on the was again to understand this (probably), but is it correct to conclude that practically no delay introduces no microstep change whereas zero delay does?

Reply:
> I need to bang me head several times on the was again to understand this (probably), but is it correct to conclude that practically no delay introduces no microstep change whereas zero delay does?

This is also what I'm unsure about, actually. If we treat -:infinity: as zero, then adding it to some other tag should yield the same result as adding zero to it. This suggests that we should have
```
getLaterTag(delay: TimeValue): Tag { 
  if (this.time.isNever() || delay.isNever() || delay.isZero()) { 
    return new Tag(this.time, this.microstep + 1);
  if (this.time.isForever() || delay.isForever()) {
    return forever();
  } else { 
    return new Tag(delay.add(this.time), 0); 
  }
} 
```

WDYT, @edwardalee?

Reply:
I have a concern about the comment
>then adding it to some other tag should yield the same result as adding zero to it.

Because, @edwardalee said that
>we have to make a distinction between zero delay and no delay.

More concretely, when a federate sends two messages at the tag (0, 0) with `NEVER` and `ZERO` additional delays, the destination federate schedules them at the tag (0, 0) and (0, 1) respectively. Don't we have to distinguish those no and zero delays?

Reply:
I think I'm getting it now, but I think we're conflating two issues.

1. How do we determine the tags of events sent from one federate to another (this depends on the delay along the connection). 
2. How does the arithmetic of tags work.

As for 1, I see no reason why we need to encode "no delay" as `NEVER`, as is done in the C implementation; in TypeScript, we can just record it as `undefined` or `null`.

As for 2, my view is expressed in https://github.com/lf-lang/reactor-ts/pull/186/files#r1275558491

Reply:
'no delay' is expressed by `undefined` in [26d0a9f](https://github.com/lf-lang/reactor-ts/pull/186/commits/26d0a9f240c3619d9d8b899d163697047116bd7c). Also, `NEVER` and `FOREVER` is treated as -♾️ and ♾️, unlike `reactor-c`. 
I left a FIXME in case we will add a microstep when we add `ZERO` to a tag `t` like `reactor-c`. 
https://github.com/lf-lang/reactor-ts/blob/2b36efb609c67f29bee1fedfbefa2b280424f253/src/core/time.ts#L470-L477

Reply:
Marking this as resolved.